### ISSUE TITLE
[Bug] FlexLayout does not shrink when last row of children are hidden

### ISSUE THREAD

**Issue by ssombat** [xamarin/Xamarin.Forms#5959]
### Description
Flexlayout should shrink when its children are hidden. When hiding children at the last row, flexlayout will no longer shrink.
### Steps to Reproduce
Using my project that I provided
1. Click one of the white boxes that are on the last row. (That white box is now hidden)
2. Click on any of the white boxes to hide them. (repeat this step as many times as you like)
3. Flexlayout (blue background) does not shrink.

Optional Steps (to see intended behavior)
1. Restart the app or hit the reset button if the above steps were preformed. 
2. Click on any of the white boxes that are NOT in the last row to hide them. (repeat this step as many times as you like)
3. Flexlayout (blue background) will start to shrink until there is only one row left.
### Expected Behavior
As children are being hidden regardless of its position, flexlayout should shrink if there is unused space.
### Actual Behavior
Hiding children in the last row, will cause flexlayout to not shrink.
### Basic Information

- Version with issue:  3.6.0.264807
- In UWP and Android

### Screenshots
When children in the last row are hidden
![image](https://user-images.githubusercontent.com/25404238/56426601-e0e8be80-6286-11e9-84b9-35ec670dd109.png)

When children NOT in the last row are hidden
![image](https://user-images.githubusercontent.com/25404238/56426671-24432d00-6287-11e9-8367-a1bc590da56b.png)

### Reproduction Link
https://github.com/ssombat/Xamarin-Flexlayout-Issue/blob/master/App1.zip

**Comment by Avitose** [comment]
Bug still persists in Xamarin Froms 5.0.0.2244

**Comment by Miksier** [comment]
So basically its very easy to fix the symptoms in the XF SC, in measuring we just have to add check for visibility. But it doesnt solve the root cause of the issue. Maybe when I find it I'll create a PR with a fix.



For the  mean time I found a way to hack away this bug. 

You create a new class that derives from FlexLayout and add this code:


`private bool _useOvverridenLogic = false;

		protected override SizeRequest OnMeasure(double widthConstraint, double heightConstraint)
		{
			var baseMeasure = base.OnMeasure(widthConstraint, heightConstraint).Request;

			if (_useOvverridenLogic)
			{
				heightConstraint = 0;
				foreach (var item in Children)
				{
					if (item.IsVisible)
					{
						heightConstraint = Math.Max(heightConstraint, item.Y + item.Height + item.Margin.Bottom);
					}
				}
				return new SizeRequest(new Size(baseMeasure.Width, heightConstraint));
			}

			return new SizeRequest(baseMeasure);
		}

		protected override void OnSizeAllocated(double width, double height)
		{
			base.OnSizeAllocated(width, height);
			if (width != -1 && height != -1 && !_useOvverridenLogic )
			{
				_useOvverridenLogic = true;
				InvalidateMeasure();
			}
		}`
Basically we calculate the height of flex ourselves and since the first time we dont know the height of children we use OnSizeAllocated. When size is alocated the size of children is known so we can correctly calculate it.
It'd be different and more optimized if we had access to the _root field and private class Flex.Item. But we dont so we have to use what we have.

**Comment by lanatias** [comment]
I have tested this with Xamarin.Forms 5.0.0.2291 and was able to reproduce the issue as well. 

<img width="178" alt="Screenshot 2022-01-26 224248" src="https://user-images.githubusercontent.com/5842621/151244380-fa8fa664-5cba-43a1-9ac8-5dcf20083370.png">
<img width="179" alt="Screenshot 2022-01-26 224311" src="https://user-images.githubusercontent.com/5842621/151244372-760e96bf-bd15-435b-9772-206975672229.png">
<img width="179" alt="Screenshot 2022-01-26 224219" src="https://user-images.githubusercontent.com/5842621/151244378-6dad63d9-79a5-404b-99af-e61b88997725.png">

Initial render works fine (display the page with no empty spaces), but as soon as I toggled between Categories (change the visibility of FlexLayout items), I can see the empty spaces. 

I have attached my test project below to reproduce the above issues:

[FlexLayoutTest.zip](https://github.com/xamarin/Xamarin.Forms/files/7945475/FlexLayoutTest.zip)

**Comment by raver99** [comment]
Confirming that it is still happening in Xamarin.Forms 5.0.0.2291

@[Pastajello](https://github.com/Pastajello) was a quick fix. Thanks
